frontend/src/metabase/visualizations/components/settings/ChartSettingColumnEditor.tsx

```diff
@@ -0,0 +1,204 @@
+          .filter(isNotNull),
+        fks: [],
+      };
+    } else {
+      return {
+        count: 0,
+        dimensions: [],
+        fks: [],
+      };
+    }
+  }, [question, columns]);
+
+  const getColumnSettingByDimension = (dimension: Dimension) => {
+    return columnSettings.find(setting =>
+      dimension.isSameBaseDimension(setting.fieldRef as ConcreteField),
+    );
+  };
+
+  const columnIsEnabled = (dimension: Dimension) => {
+    return getColumnSettingByDimension(dimension)?.enabled;
+  };
+
+  const toggleColumn = (dimension: Dimension) => {
+    const setting = getColumnSettingByDimension(dimension);
+
+    if (!setting?.enabled) {
+      enableColumns([dimension]);
+    } else {
+      disableColumns([dimension]);
+    }
+  };
+
+  const tableInColumnSettings = (dimensions: Dimension[]) => {
+    return dimensions.some(
+      dimension => getColumnSettingByDimension(dimension)?.enabled,
+    );
+  };
+
+  const disableColumns = (dimensions: Dimension[]) => {
+    const columnSettingsCopy = columnSettings.map(setting => ({
+      ...setting,
+      enabled: dimensions.some(dimension =>
+        dimension.isSameBaseDimension(setting.fieldRef as ConcreteField),
+      )
+        ? false
+        : setting.enabled,
+    }));
+
+    onChange(columnSettingsCopy);
+  };
+
+  const enableColumns = (dimensions: Dimension[]) => {
+    const [dimensionsInColumnSettings, dimensionsNotInColumnSettings] =
+      _.partition(
+        dimensions,
+        dimension => !!getColumnSettingByDimension(dimension),
+      );
+
+    //Enable any columns that are in column settings (used for Native Queries)
+    const columnSettingsCopy = columnSettings.map(setting => ({
+      ...setting,
+      enabled: dimensionsInColumnSettings.some(dimension =>
+        dimension.isSameBaseDimension(setting.fieldRef as ConcreteField),
+      )
+        ? true
+        : setting.enabled,
+    }));
+
+    //Add any that are not in the ColumnSettings (used for Structured Queries)
+    onChange([
+      ...columnSettingsCopy,
+      ...dimensionsNotInColumnSettings.map(dimension => ({
+        fieldRef: dimension.mbql(),
+        enabled: true,
+      })),
+    ]);
+  };
+
+  return (
+    <div className="list">
+      <TableHeaderContainer>
+        <TableName>{fieldOptions.name}</TableName>
+        <BulkActionButton
+          tableInColumns={tableInColumnSettings(fieldOptions.dimensions)}
+          bulkEnable={() => enableColumns(fieldOptions.dimensions)}
+          bulkDisable={() => disableColumns(fieldOptions.dimensions)}
+        />
+      </TableHeaderContainer>
+      {fieldOptions.dimensions.map((dimension, index) => (
+        <FieldCheckbox
+          label={dimension.displayName()}
+          onClick={() => toggleColumn(dimension)}
+          checked={columnIsEnabled(dimension)}
+          key={`${dimension.displayName()}-${index}`}
+        />
+      ))}
+      {fieldOptions.fks.length > 0 &&
```

Comment:
Do we need this check for length before calling the map?

Reply:
You're right, I don't need to haha. This evolved and now it's no longer necessary